File: shop/src/main/java/basic/shop/order/command/domain/value/OrderLine.java
package basic.shop.order.command.domain.value;

import basic.shop.catalog.command.domain.Product;
import basic.shop.catalog.command.domain.ProductId;
import basic.shop.common.model.AbstractBaseEntity;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.ToString;

import javax.persistence.EmbeddedId;
import javax.persistence.Entity;
import java.util.Objects;

@Entity
@Getter
@AllArgsConstructor
@ToString(of = {"id", "price", "quantity", "amounts", "product"})
public class OrderLine extends AbstractBaseEntity {
    @EmbeddedId
    private ProductId id;
    private int price;
    private int quantity;
    private int amounts;
    private Product product;

    private int calculateAmounts() {
        return price * quantity;
    }

    @Override
    public boolean equals(Object o) {
        if(o == this)
            return true;
        if(!(o instanceof OrderLine))
            return false;
        OrderLine orderLine = (OrderLine) o;
        return Objects.equals(id, orderLine.id) && Objects.equals(price, orderLine.price)
                && Objects.equals(quantity, orderLine.quantity) && Objects.equals(amounts, orderLine.amounts)
                && Objects.equals(product, orderLine.product);
    }

    @Override
    public int hashCode() {
        int result = id == null ? 0 : id.hashCode();
        result = result * 31 + Integer.hashCode(price);
        result = result * 31 + Integer.hashCode(quantity);
        result = result * 31 + Integer.hashCode(quantity);
        result = result * 31 + Integer.hashCode(amounts);
        result = result * 31 + Objects.hashCode(product);
        return result;
    }
}

File: spring-mvc-with-kotlin/order-api/src/main/kotlin/basic/orderapi/order/command/domain/value/OrderLine.kt
package basic.orderapi.order.command.domain.value

import basic.orderapi.catalog.command.domain.product.value.ProductId
import basic.orderapi.common.jpa.MoneyConverter
import basic.orderapi.common.model.Money
import javax.persistence.Convert
import javax.persistence.Embeddable
import javax.persistence.Embedded

@Embeddable
class OrderLine (
    @Embedded
    var productId: ProductId,
    @Convert(converter = MoneyConverter::class)
    var price: Money,
    var quantity: Int,
) {
    @Convert(converter = MoneyConverter::class)
    var amounts: Money = calculateAmounts(price, quantity)

    private fun calculateAmounts(price:Money, quantity: Int): Money {
        return price.multiply(quantity)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as OrderLine

        if (price != other.price) return false
        if (quantity != other.quantity) return false
        if (amounts != other.amounts) return false

        return true
    }

    override fun hashCode(): Int {
        var result = price.hashCode()
        result = 31 * result + quantity
        result = 31 * result + amounts.hashCode()
        return result
    }


}
